src/Analyzers/Core/Analyzers/AbstractBuiltInCodeStyleDiagnosticAnalyzer_Core.cs

```diff
@@ -69,12 +77,191 @@ protected AbstractBuiltInCodeStyleDiagnosticAnalyzer(ImmutableArray<DiagnosticDe
+
+        protected bool ShouldSkipAnalysis(
+            SyntaxTree tree,
+            AnalyzerOptions analyzerOptions,
+            CompilationOptions compilationOptions,
+            NotificationOption2? notification,
+            CancellationToken cancellationToken)
+            => ShouldSkipAnalysis(tree, analyzerOptions, compilationOptions, notification, performDescriptorsCheck: true, cancellationToken);
+
+        protected bool ShouldSkipAnalysis(
+            SyntaxTree tree,
+            AnalyzerOptions analyzerOptions,
+            CompilationOptions compilationOptions,
+            ImmutableArray<NotificationOption2> notifications,
+            CancellationToken cancellationToken)
+        {
+            // We need to check if the analyzer's severity has been escalated either via 'option_name = option_value:severity'
+            // setting or 'dotnet_diagnostic.RuleId.severity = severity'.
+            // For the former, we check if any of the given notifications have been escalated via the ':severity' such
+            // that analysis cannot be skipped. For the latter, we perform descriptor-based checks.
+            // Descriptors check verifies if any of the diagnostic IDs reported by this analyzer
+            // have been escalated to a severity that they must be executed.
+
+            // PERF: Execute the descriptors check only once for the analyzer, not once per each notification option.
+            var performDescriptorsCheck = true;
+
+            // Check if any of the notifications are enabled, if so we need to execute analysis.
+            foreach (var notification in notifications)
+            {
+                if (!ShouldSkipAnalysis(tree, analyzerOptions, compilationOptions, notification, performDescriptorsCheck, cancellationToken))
+                    return false;
+
+                if (performDescriptorsCheck)
+                    performDescriptorsCheck = false;
+            }
+
+            return true;
+        }
+
+        private bool ShouldSkipAnalysis(
+            SyntaxTree tree,
+            AnalyzerOptions analyzerOptions,
+            CompilationOptions compilationOptions,
+            NotificationOption2? notification,
+            bool performDescriptorsCheck,
+            CancellationToken cancellationToken)
+        {
+            // We need to check if the analyzer's severity has been escalated either via 'option_name = option_value:severity'
+            // setting or 'dotnet_diagnostic.RuleId.severity = severity'.
+            // For the former, we check if the given notification have been escalated via the ':severity' such
+            // that analysis cannot be skipped. For the latter, we perform descriptor-based checks.
+            // Descriptors check verifies if any of the diagnostic IDs reported by this analyzer
+            // have been escalated to a severity that they must be executed.
+
```

Comment:
This method has the core pre-analysis check to bail out early based on effective severity of all diagnostics that can be reported by the analyzer. It builds on top of the compiler public API added in #70131.

If the analyzer is running in the IDE host ('_minimumReportedSeverity' is Hidden), this method always returns `false`. Otherwise, if running from the command line build ('_minimumReportedSeverity' is Warning), we check if any diagnostic ID that it reports can be a Warning or an Error in the entire compilation - if not, we return `true` to avoid redundant analysis.